llmfoundry/data/finetuning/tasks.py

```diff
@@ -1033,7 +1034,16 @@ def build_from_streaming(
         *args: Any,
         **kwargs: Any,
     ) -> StreamingFinetuningDataset:
-        return self.streaming_dataset_class(*args, **kwargs)
+        dataset = self.streaming_dataset_class(*args, **kwargs)
```

Comment:
iiuc, Streaming will throw an exception here so the code below won't even run

Reply:
The error trace that @jjanezhang sent showed it erroring during data loading

Reply:
Maybe num_canonical_nodes < num_samples isn't actually the right check?

```
  File "/usr/lib/python3/dist-packages/torch/utils/data/_utils/worker.py", line 308, in _worker_loop
    data = fetcher.fetch(index)  # type: ignore[possibly-undefined]
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/torch/utils/data/_utils/fetch.py", line 32, in fetch
    data.append(next(self.dataset_iter))
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/streaming/base/dataset.py", line 1501, in __iter__
    sample_ids = self._get_work(epoch, sample_in_epoch)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/streaming/base/dataset.py", line 1036, in _get_work
    epoch_sample_ids = generate_work(self.batching_method, self, p_world, epoch,
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/streaming/base/batching/__init__.py", line 45, in generate_work
    return get(dataset, world, epoch, sample_in_epoch)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/streaming/base/batching/random.py", line 67, in generate_work_random_batching
    shuffle = get_shuffle(dataset.shuffle_algo, shuffle_units, dataset.num_canonical_nodes,
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/streaming/base/shuffle/__init__.py", line 50, in get_shuffle
    return get(shard_sizes, num_canonical_nodes, seed, epoch, block_size)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/streaming/base/shuffle/py1e.py", line 76, in get_shuffle_py1e
    raise ValueError(f'The number of samples assigned to a canonical node is 0. This ' +
[2024-09-10 08:40:44.473000+00:00][48225]: ValueError: The number of samples assigned to a canonical node is 0. This very likely indicates that the number of samples in this stream is less than the number of canonical nodes, which is 4. Please check your index.json file and ensure that your dataset has been written out correctly. If this was intended, reduce num_canonical_nodes.

```

Reply:
Checked and streaming does not throw an error. I also realized streaming doesn't initialize the number of canonical nodes (if it's None) until dataloading starts, so I added a check to make sure the number of samples at least the number of physical nodes.